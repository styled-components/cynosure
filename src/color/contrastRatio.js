// @flow
import getLuminance from './getLuminance'
import curry from '../internalHelpers/_curry'

/**
 * Given two colors, returns the contrast ratio.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   background: contrastRatio('#CCCD64', '#0000ff') > 0.5 ? '#CCCD64' : '#0000ff',
 *   background: contrastRatio('rgba(58, 133, 255, 1)', 'rgba(255, 57, 149, 1)') > 0.5 ?
 *                             'rgba(58, 133, 255, 1)' :
 *                             'rgba(255, 57, 149, 1)',
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   background: ${contrastRatio('#CCCD64', '#0000ff') > 0.5 ? '#CCCD64' : '#0000ff'};
 *   background: ${contrastRatio('rgba(58, 133, 255, 1)', 'rgba(255, 57, 149, 1)') > 0.5 ?
 *                             'rgba(58, 133, 255, 1)' :
 *                             'rgba(255, 57, 149, 1)'};
 *
 * // CSS in JS Output
 *
 * div {
 *   background: "#CCCD64";
 *   background: "rgba(58, 133, 255, 1)";
 * }
 */

function contrastRatio(color1: string, color2: string): number {
  return Math.abs(getLuminance(color1) - getLuminance(color2))
}

const curriedContrastRatio = curry(contrastRatio)
export default curriedContrastRatio
