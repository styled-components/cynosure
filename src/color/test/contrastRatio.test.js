// @flow
import contrastRatio from '../contrastRatio'

describe('contrastRatio', () => {
  it('should return the contrast ratio given white (#fff) and black (#000)', () => {
    expect(contrastRatio('#fff', '#000')).toMatchSnapshot()
  })

  it('should return the contrast ratio given black (#000) and white (#fff)', () => {
    expect(contrastRatio('#000', '#fff')).toMatchSnapshot()
  })

  it('should return the contrast ratio given white (#fff) and white (#fff)', () => {
    expect(contrastRatio('#fff', '#fff')).toMatchSnapshot()
  })

  it('should return the contrast ratio given black (#000) and black (#000)', () => {
    expect(contrastRatio('#000', '#000')).toMatchSnapshot()
  })

  it('should return the contrast ratio between two hex colors', () => {
    expect(contrastRatio('#123', '#456')).toMatchSnapshot()
  })

  it('should return the contrast ratio between twon rgba colors', () => {
    expect(contrastRatio('rgba(204,205,100,0.7)', 'rgba(100,0,255,0.3)')).toMatchSnapshot()
  })

  it('should return the contrast ratio between twon rgb colors', () => {
    expect(contrastRatio('rgb(204,0,200)', 'rgb(0,205,0)')).toMatchSnapshot()
  })

  it('should return the contrast ratio between twon hlsa colors', () => {
    expect(contrastRatio('hsla(250, 100%, 50%, 0.2)', 'hsla(20, 50%, 100%, 0.5)')).toMatchSnapshot()
  })

  it('should return the contrast ratio between twon hls colors', () => {
    expect(contrastRatio('hsl(0, 100%, 50%)', 'hsl(100, 20%, 80%)')).toMatchSnapshot()
  })

  it('should return the contrast ratio between two named CSS colors', () => {
    expect(contrastRatio('papayawhip', 'blue')).toMatchSnapshot()
  })
})
