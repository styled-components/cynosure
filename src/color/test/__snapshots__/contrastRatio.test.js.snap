// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contrastRatio should return the contrast ratio between two hex colors 1`] = `0.07183009969244383`;

exports[`contrastRatio should return the contrast ratio between two named CSS colors 1`] = `0.8057710019983539`;

exports[`contrastRatio should return the contrast ratio between twon hls colors 1`] = `0.4314568366064737`;

exports[`contrastRatio should return the contrast ratio between twon hlsa colors 1`] = `0.9226640873414479`;

exports[`contrastRatio should return the contrast ratio between twon rgb colors 1`] = `0.26655143220211924`;

exports[`contrastRatio should return the contrast ratio between twon rgba colors 1`] = `0.4749078741492648`;

exports[`contrastRatio should return the contrast ratio given black (#000) and black (#000) 1`] = `0`;

exports[`contrastRatio should return the contrast ratio given black (#000) and white (#fff) 1`] = `1`;

exports[`contrastRatio should return the contrast ratio given white (#fff) and black (#000) 1`] = `1`;

exports[`contrastRatio should return the contrast ratio given white (#fff) and white (#fff) 1`] = `0`;
